test/semantics/misc-declarations.f90

```diff
@@ -0,0 +1,49 @@
+! Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+!
+! Licensed under the Apache License, Version 2.0 (the "License");
+! you may not use this file except in compliance with the License.
+! You may obtain a copy of the License at
+!
+!     http://www.apache.org/licenses/LICENSE-2.0
+!
+! Unless required by applicable law or agreed to in writing, software
+! distributed under the License is distributed on an "AS IS" BASIS,
+! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+! See the License for the specific language governing permissions and
+! limitations under the License.
+
+! Miscellaneous constraint and requirement checking on declarations:
+! - 8.5.19 constraints on the VOLATILE attribute
+
+module m
+  type :: hasCoarray
+    real :: coarray[:]
+  end type
+  real :: coarray[:]
```

Comment:
I believe both coarrays need to be allocatable or have explicit cobounds.

Reply:
Right, that colon should be an asterisk.  And we should check C827 & C828.  Will do both.

Reply:
Adding checks for these constraints caught a whole load of additional incorrect usage in other tests, so thanks for noticing.